src/pd/pd.rs

```diff
@@ -107,14 +110,114 @@ impl Default for StoreStat {
+    down_peers: Vec<pdpb::PeerStats>,
+    pending_peers: Vec<Peer>,
+    approximate_size: u64,
+
+    read_bytes_delta: u64,
+    read_keys_delta: u64,
+    written_bytes_delta: u64,
+    written_keys_delta: u64,
+
+    last_written_bytes: u64,
+    last_written_keys: u64,
+
+    valid: bool,
+}
+
+impl RegionHeartbeatCache {
+    fn heartbeat(&mut self) -> Option<(Region, Peer, RegionStat)> {
+        if self.valid {
+            return None;
+        }
+        // Somthing have changed, send the heartbeat to PD.
+        self.valid = true;
+        let region_stat = RegionStat {
+            down_peers: self.down_peers.clone(),
+            pending_peers: self.pending_peers.clone(),
+            written_bytes: self.written_bytes_delta,
+            written_keys: self.written_keys_delta,
+            read_bytes: self.read_bytes_delta,
+            read_keys: self.read_keys_delta,
+            approximate_size: self.approximate_size,
+        };
+        self.read_bytes_delta = 0;
+        self.read_keys_delta = 0;
+        self.written_bytes_delta = 0;
+        self.written_keys_delta = 0;
+        Some((self.region.clone(), self.peer.clone(), region_stat))
+    }
+
+    fn merge_flow_statistics(&mut self, stats: FlowStatistics) {
+        if stats.read_bytes > 0 {
+            self.read_bytes_delta += stats.read_bytes as u64;
+            self.valid = false;
+        }
+        if stats.read_keys > 0 {
+            self.read_keys_delta += stats.read_keys as u64;
+            self.valid = false;
+        }
+    }
+
+    fn merge_heartbeat(&mut self, db: &DB, hb_task: Task) {
+        match hb_task {
+            Task::Heartbeat {
+                region,
+                peer,
+                down_peers,
+                pending_peers,
+                total_written_bytes,
+                total_written_keys,
+                region_size,
+            } => {
+                let approximate_size = match region_size {
+                    Some(size) => size,
+                    None => get_region_approximate_size(db, &region).unwrap_or(0),
+                };
+                if self.approximate_size != approximate_size {
+                    self.approximate_size = approximate_size;
+                    self.valid = false;
+                }
+
+                if self.region != region {
+                    self.region = region;
+                    self.valid = false;
+                }
+
+                if self.peer != peer {
+                    self.peer = peer;
+                    self.valid = false;
+                }
+
+                if self.down_peers != down_peers {
+                    self.down_peers = down_peers;
+                    self.valid = false;
+                }
+
+                if self.pending_peers != pending_peers {
+                    self.pending_peers = pending_peers;
+                    self.valid = false;
+                }
+
+                if self.last_written_bytes < total_written_bytes {
```

Comment:
I think `Task::Heartbeat` needs a new field `become_leader` to indicates the it's a new leader or not. Or suppose this case for leader on store X:
> transfer leader to Y;
> remove-peer from X;
> add-new-peer on X;
> transfer leader to X again.

`total_written_bytes` maybe less than `last_written_bytes`. And it's not coverd here.

Reply:
`total_written_bytes` is monotonically increasing even if it spilt or changed role.